did-tezos/src/lib.rs

```diff
@@ -1359,6 +1375,49 @@ mod tests {
         // assert_eq!(d, expected);
     }
 
+    #[tokio::test]
+    async fn test_full_resolution_kt1() {
+        let live_did_manager = format!("did:tz:{}", LIVE_KT1);
+
+        let (res_meta, res_doc, _res_doc_meta) = DIDTZ
+            .resolve(&live_did_manager, &ResolutionInputMetadata::default())
+            .await;
+        assert_eq!(res_meta.error, None);
+        let d = res_doc.unwrap();
+        let expected = Document {
+            id: live_did_manager.clone(),
+            verification_method: Some(vec![
+                VerificationMethod::Map(VerificationMethodMap {
+                    id: format!("{}#blockchainAccountId", live_did_manager),
+                    type_: "Ed25519PublicKeyBLAKE2BDigestSize20Base58CheckEncoded2021".to_string(),
+                    blockchain_account_id: Some(format!("tezos:{}:{}", LIVE_NETWORK, LIVE_KT1)),
+                    controller: live_did_manager.clone(),
+                    property_set: None,
+                    ..Default::default()
+                }),
+                VerificationMethod::DIDURL(DIDURL {
+                    did: format!("did:pkh:tz:{}", LIVE_TZ1),
+                    path_abempty: "".to_string(),
+                    query: None,
+                    fragment: Some("TezosMethod2021".to_string()),
+                }),
+            ]),
```

Comment:
To date we've been using public key hashes in the `blockchainAccountId` property. The hashing for verification is implemented in `src/caip10.rs`. A contract address is valid for CAIP-10, and therefore should be valid for blockchainAccountId, but it doesn't fit into the pattern of hashing a public key to compare with an account address for verification purposes. So I don't think we could use this verification method in ssi currently, without adding support for resolving the contract address to a public key or public key hash in ssi, e.g. by calling `getAuthentication` on the contract.

Using `did:pkh` as in the second verification method URL, I think makes sense, because it refers unambiguously to the public key (compared to using did:tz:tz where there is a contract lookup involved - unless it is desired to enable recursive key rotation). did:pkh also uses CAIP-10, but having the method name meaning "public key hash" I think implies that it's only intended for use with account addresses, not contract addresses (although that should be confirmed in the did:pkh spec). It should also probably be updated in the did:tz spec to use did:pkh in this way - while considering backwards compatibility.

did:pkh is deprecating `did:pkh:tz:` in favor of `did:pkh:tezos:<chain>:` to follow CAIP-10 - although the chain id is not used for verification purposes here.

I think TezosMethod2021 can be deprecated, as the proof type it goes with (TezosSignature2021) can just use the default verification method. Since the did:tz:KT1 implementation here is not yet merged, I don't think it's in use, so I would suggest removing TezosMethod2021 here so it doesn't need to be kept for any legacy reasons.

With the verification methods using did:pkh, I don't think they need to be included in the verificationMethod array in the did:tz:KT1 DID document - just the URLs can be listed in `authentication`, `assertionMethod`, etc. As the did:pkh DID document will need to be resolved anyway, to verify the controller property of the verification method object, I don't think there would be much benefit including the cross-DID-method verification method object inline in this DID document - unless we either want to enable skipping verification of the controller property (effectively to trust the did:tz implementation's did:pkh resolution, or in case there is no separate did:pkh resolver available) or add some additional verification (to require that the did:tz implementation's did:pkh implementation agrees with a separate did:pkh implementation).

Potential complexity of DID method dependency (tz -> pkh) could be avoided by extending did:tz with a new DID subtype (or using DID paths) for unrotateable account addresses - i.e. putting did:pkh:tz into did:tz: somehow.

```suggestion
            authentication: Some(vec![
                VerificationMethod::DIDURL(DIDURL {
                    did: format!("did:pkh:tezos:{}:{}", LIVE_NETWORK, LIVE_TZ1),
                    path_abempty: "".to_string(),
                    query: None,
                    fragment: Some("blockchainAccountId".to_string()),
                }),
            ]),
```

This change I am suggesting is based on resolving the current DID with DIDKit using ssi @ 92e2f9c0efdfa0b6974d114814c8530c074344b6 from this PR:
```js
$ didkit did-resolve did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi
{
  "@context": [
    "https://www.w3.org/ns/did/v1",
    {
      "Ed25519PublicKeyBLAKE2BDigestSize20Base58CheckEncoded2021": "https://w3id.org/security#Ed25519PublicKeyBLAKE2BDigestSize20Base58CheckEncoded2021",
      "blockchainAccountId": "https://w3id.org/security#blockchainAccountId"
    }
  ],
  "id": "did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi",
  "verificationMethod": [
    {
      "id": "did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi#blockchainAccountId",
      "type": "Ed25519PublicKeyBLAKE2BDigestSize20Base58CheckEncoded2021",
      "controller": "did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi",
      "blockchainAccountId": "tezos:NetXdQprcVkpaWU:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi"
    },
    "did:pkh:tz:tz1giDGsifWB9q9siekCKQaJKrmC9da5M43J#TezosMethod2021"
  ],
  "authentication": [
    "did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi#blockchainAccountId"
  ],
  "assertionMethod": [
    "did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi#blockchainAccountId"
  ],
  "service": [
    {
      "id": "did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi#discovery",
      "type": "TezosDiscoveryService",
      "serviceEndpoint": "http://example.com"
    }
  ]
}
```
I'm suggesting it should result in the following instead:
```json
{
  "@context": [
    "https://www.w3.org/ns/did/v1",
    {
      "Ed25519PublicKeyBLAKE2BDigestSize20Base58CheckEncoded2021": "https://w3id.org/security#Ed25519PublicKeyBLAKE2BDigestSize20Base58CheckEncoded2021",
      "blockchainAccountId": "https://w3id.org/security#blockchainAccountId"
    }
  ],
  "id": "did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi",
  "authentication": [
    "did:pkh:tezos:NetXdQprcVkpaWU:tz1giDGsifWB9q9siekCKQaJKrmC9da5M43J#blockchainAccountId"
  ],
  "assertionMethod": [
    "did:pkh:tezos:NetXdQprcVkpaWU:tz1giDGsifWB9q9siekCKQaJKrmC9da5M43J#blockchainAccountId"
  ],
  "service": [
    {
      "id": "did:tz:KT1ACXxefCq3zVG9cth4whZqS1XYK9Qsn8Gi#discovery",
      "type": "TezosDiscoveryService",
      "serviceEndpoint": "http://example.com"
    }
  ]
}
```

For a similar example of existing cross-DID-method use, here is did:web with `did:key` - this works with ssi/DIDKit, although other VC API Test Suite implementers don't seem to support it. VC:
```json
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1"
  ],
  "id": "https://demo.didkit.dev/2021/12/03/vc.jsonld",
  "type": [
    "VerifiableCredential"
  ],
  "credentialSubject": {},
  "issuer": "did:web:demo.didkit.dev:2021:12:03",
  "issuanceDate": "2021-12-03T20:18:09Z",
  "proof": {
    "type": "Ed25519Signature2018",
    "proofPurpose": "assertionMethod",
    "verificationMethod": "did:key:z6MknzD3XyJNXWBjHTKxp8HqtsRnbazPRowYcSzHFuRF74B5#z6MknzD3XyJNXWBjHTKxp8HqtsRnbazPRowYcSzHFuRF74B5",
    "created": "2021-12-03T20:18:09.556Z",
    "jws": "eyJhbGciOiJFZERTQSIsImNyaXQiOlsiYjY0Il0sImI2NCI6ZmFsc2V9..lssWnaAid-cCl0SM_d7BYhouQ9lsyis3Mp6CBfE4q8mL8PNuQ2AuoF-cPDJ9bGPdgoHiCAxLYikmz2aiaaU4Bw"
  }
}
```
Issuer:
```json
{
  "@context": "https://www.w3.org/ns/did/v1",
  "id": "did:web:demo.didkit.dev:2021:12:03",
  "authentication": [
    "did:key:z6MknzD3XyJNXWBjHTKxp8HqtsRnbazPRowYcSzHFuRF74B5#z6MknzD3XyJNXWBjHTKxp8HqtsRnbazPRowYcSzHFuRF74B5"
  ],
  "assertionMethod": [
    "did:key:z6MknzD3XyJNXWBjHTKxp8HqtsRnbazPRowYcSzHFuRF74B5#z6MknzD3XyJNXWBjHTKxp8HqtsRnbazPRowYcSzHFuRF74B5"
  ]
}
```

Reply:
Sooo atm the VM stored in the contract it the full `did:pkh:tz:tz1giDGsifWB9q9siekCKQaJKrmC9da5M43J#TezosMethod2021`. Do you think that was a bad idea and should have only been the tz/kt address?

Reply:
I don't know; but I think that simplifies what did:tz:KT1 has to do: if the contract contains a fully-qualified verification method DID URL, that could be passed through as a string in the `authentication` array (and/or other verification relationships that did:tz will support)

Reply:
Frankly I'm lost, am I on the right track with 7cb2cb4? There are still a couple of tests failing due to proof filtering

Reply:
As mentioned in https://github.com/spruceid/ssi/pull/363#issuecomment-993624124, okay as-is (57f5e3d reverting 7cb2cb4); probably should update spec before making more significant changes.